File: sorting/python/radixsort.py
import unittest


def radix_sort(A):
    # create an array with
    Bucket = []
    # create ten arrays inside Bucket
    for i in range(10):
        Bucket.append([])

    for i in range(1, 5):
        for k in range(len(A)):
            Bucket[get_digit_value(A[k], i)].append(A[k])
        # sort based on Bucket content
        pour_bucket_content_to_array(Bucket, A)
    return A


def pour_bucket_content_to_array(Bucket, A):
    counter = 0
    for i in range(len(Bucket)):
        for k in range(len(Bucket[i])):
            A[counter] = Bucket[i][k]
            counter += 1
        # empty Bucket i
        del Bucket[i][:]


def get_digit_value(num, place):
    return (num / (10 ** (place - 1)) % 10)


class UnitTestClass(unittest.TestCase):

    def test_radix_sort(self):
		self.assertEquals(radix_sort([5231, 4231, 414, 1004]), [414, 1004, 4231, 5231])
		self.assertEquals(radix_sort([6140, 2010, 3000, 5]), [5, 2010, 3000, 6140])
		self.assertEquals(radix_sort([1234, 1, 2, 3]), [1, 2, 3, 1234])

    def test_get_digit_value(self):
        self.assertEquals(get_digit_value(54321, 3), 3)

if __name__ == "__main__":
    unittest.main()

File: sorting/java/src/RadixSort.java
import java.util.*;
import java.io.File;
import java.util.Arrays;

public class RadixSort extends SortingAlgorithm {
	private ArrayList<ArrayList> Bucket;
	public RadixSort( int[] arr, int n)
	{
		super( arr, n ); 
	}

	public void sort()
	{
		// get the maximum value; then get its number of digits
		int maxDigitCount = 5; // this for now

		// initialize the Bucket; an array that contains ArrayList
		Bucket = new ArrayList<ArrayList>();
		for(int i=0; i < 10; i++)
			Bucket.add(new ArrayList());

		for(int i=1; i < maxDigitCount; i++)
		{
			for(int k=0; k < n; k++)
			{
				Bucket.get(getDigitValue(arr[k], i)).add(arr[k]);
			}
			pourBucketsContentToArray();
		}
	}

	/**
	 *	Set the values inside the Bucket to an array 
	 * 	then it empties the bucket after
	 */
	private void pourBucketsContentToArray() 
	{
		int counter = 0;
		// System.out.println(Bucket);
		for(int i=0; i<10; i++)
		{
			for(int k=0; k < Bucket.get(i).size(); k++)
			{
				arr[counter] = (int) Bucket.get(i).get(k);
				counter++;
			}

			// clear each Bucket
			Bucket.get(i).clear();
		}
	}

	/**
	 * Returns the digit in the nth place
	 */
	private int getDigitValue(int num, int place)
	{
		int d = ((int) (num / Math.pow(10, place-1))) % 10;
		return d;
	}
	
}